### ISSUE TITLE
复现结果数据问题

### ISSUE THREAD

**Issue by ZMingzhu** [ACAT-SCUT/CycleNet#12]
作者大大您好！我在复现本文代码的时候发现了一些问题：
（1）我在没有更改代码参数和结构的情况下直接运行“sh run_main.sh”语句复现ETTH数据集，却得到了一堆一维负数的预测结果，这是正常的吗？或者说我是哪里步骤出了问题吗，该如何解决呢？（数据结果示例在末尾）
（2）我看到代码中是有多元预测多元的参数设定（“features”参数中的M用于多元预测多元），我想要使用多参数输入预测多参数输出，但是为什么我设定之后，每次输出的结果还是只有一堆一维的负数（与问题（1）中说明的预测结果类似，一堆一维负数值），请问多输入预测多输出这样的情况该代码是否能够实现呢？我需要修改哪部分内容呢？
（3）并且在我没有修改代码的情况下，按理来说训练过程中生成的两个txt应该一个保存预测值一个保存真值，但是为什么我训练生成的两个txt结果文件中数据是一样的？为了验证这个问题产生的原因，我重新git clone了github代码仓中的原始代码重新训练测试，结果还是与之前一样（预测txt写入结果和真值txt写入结果数据一样），这个我也不是很理解哪里出了问题（示例结果在末尾，为一个训练模式下写入的预测和真值结果）。
谢谢作者大大！！

随机选择的某一训练模式下的部分预测和真值写入结果示例：
pred.txt文件中写入的部分结果示例（1-25行）：
-3.103868365287780762e-01
-2.490343302488327026e-01
-3.410085737705230713e-01
-2.720279097557067871e-01
-2.030472606420516968e-01
-1.723166108131408691e-01
-2.490343302488327026e-01
-2.260408401489257812e-01
-1.723166108131408691e-01
-1.493230164051055908e-01
-2.796560823917388916e-01
-3.486367166042327881e-01
-4.023609757423400879e-01
-3.486367166042327881e-01
-5.173286795616149902e-01
-5.939373970031738281e-01
-7.090141177177429199e-01
-8.239817619323730469e-01
-8.469753265380859375e-01
-8.929623961448669434e-01
-9.695712327957153320e-01
-1.023295521736145020e+00
-1.099904298782348633e+00
-1.130635023117065430e+00
-1.260967969894409180e+00

true.txt文件中写入的部分结果示例（1-25行）：
-3.103868365287780762e-01
-2.490343302488327026e-01
-3.410085737705230713e-01
-2.720279097557067871e-01
-2.030472606420516968e-01
-1.723166108131408691e-01
-2.490343302488327026e-01
-2.260408401489257812e-01
-1.723166108131408691e-01
-1.493230164051055908e-01
-2.796560823917388916e-01
-3.486367166042327881e-01
-4.023609757423400879e-01
-3.486367166042327881e-01
-5.173286795616149902e-01
-5.939373970031738281e-01
-7.090141177177429199e-01
-8.239817619323730469e-01
-8.469753265380859375e-01
-8.929623961448669434e-01
-9.695712327957153320e-01
-1.023295521736145020e+00
-1.099904298782348633e+00
-1.130635023117065430e+00
-1.260967969894409180e+00

**Comment by lss-1138** [comment]
gt = np.concatenate((input[0, :, -1], true[0, :, -1]), axis=0)
pd = np.concatenate((input[0, :, -1], pred[0, :, -1]), axis=0)
visual(gt, pd, os.path.join(folder_path, str(i) + '.pdf'))
np.savetxt(os.path.join(folder_path, str(i) + '.txt'), pd)
np.savetxt(os.path.join(folder_path, str(i) + 'true.txt'), gt)

---
1）数据是标准化后的，负值这段时间是低于平均值，正常
2）这部分只输出最后一维的预测结果，即使是多预测多。可以自己修改代码
3）两个文件的前半部分都是输入的X部分，而非预测的Y部分，所以是一样的

**Comment by ZMingzhu** [comment]
（1）对于数据是标准化后的，负值这段时间是低于平均值，正常：
def forward(self, x, cycle_index):
        # x: (batch_size, seq_len, enc_in), cycle_index: (batch_size,)

        # instance norm
        if self.use_revin:
            seq_mean = torch.mean(x, dim=1, keepdim=True)
            seq_var = torch.var(x, dim=1, keepdim=True) + 1e-5
            x = (x - seq_mean) / torch.sqrt(seq_var)

        # remove the cycle of the input data
        x = x - self.cycleQueue(cycle_index, self.seq_len)

        # forecasting with channel independence (parameters-sharing)
        y = self.model(x.permute(0, 2, 1)).permute(0, 2, 1)

        # add back the cycle of the output data
        y = y + self.cycleQueue((cycle_index + self.seq_len) % self.cycle_len, self.pred_len)

        # instance denorm 
        if self.use_revin:
            y = y * torch.sqrt(seq_var) + seq_mean

        return y
在CycleNet.py模型代码中，有去归一化操作，按理来说得到的应该是去归一化后的标准数值，但是按照您的说法结果还是归一化后的数据吗？
（2）对于这部分只输出最后一维的预测结果，即使是多预测多。 可以自己修改代码：
请问多预测多这种模型下，多个输入之间是会相互影响的吗？
（3）对于两个文件的前半部分都是输入的X部分，而非预测的Y部分，所以是一样的：
在测试过程中保存预测和真值结果文件，然而在原始数据集中未找到与真值文件中的数据对应的数值。

**Comment by lss-1138** [comment]
1) 模型里面的是实例归一化，此外，在Dataloader里还有个全局归一化。所以这是全局归一化后的结果
2）CycleNet是通道独立模型，可以实现多预测多，但是通道之间不会相互影响
3）参考1），保存的是全局归一化后的数据，跟原始数据不一样
